python/lsst/ctrl/mpexec/preExecInit.py

```diff
@@ -49,22 +50,50 @@ class PreExecInit:
         Data butler instance.
     taskFactory : `~lsst.pipe.base.TaskFactory`
         Task factory.
-    skipExisting : `bool`, optional
+    extendRun : `bool`, optional
         If `True` then do not try to overwrite any datasets that might exist
         in ``butler.run``; instead compare them when appropriate/possible.  If
         `False`, then any existing conflicting dataset will cause a butler
         exception to be raised.
     """
-    def __init__(self, butler, taskFactory, skipExisting=False):
+    def __init__(self, butler, taskFactory, extendRun=False):
         self.butler = butler
         self.taskFactory = taskFactory
-        self.skipExisting = skipExisting
-        if self.skipExisting and self.butler.run is None:
+        self.extendRun = extendRun
+        if self.extendRun and self.butler.run is None:
             raise RuntimeError(
-                "Cannot perform skipExisting logic unless butler is initialized "
+                "Cannot perform extendRun logic unless butler is initialized "
                 "with a default output RUN collection."
             )
 
+    @classmethod
+    def datasetTypeNames(cls, taskDefs: Iterable[TaskDef]) -> Iterator[str]:
+        """Return the names of all dataset types for a given TaskDefs.
+
+        Parameters
+        ----------
+        taskDefs : iterable of `TaskDef`
+            TaskDefs constituting a pipeline.
+
+        Yields
+        ------
+        datasetTypeName : `str`
+            Name of the dataset type.
+        """
+        # Package versions dataset type
+        yield "packages"
+
+        for taskDef in taskDefs:
+
+            # all task InitOutputs
+            for name in taskDef.connections.initOutputs:
+                attribute = getattr(taskDef.connections, name)
+                yield attribute.name
+
+            # config dataset name
```

Comment:
This method returns the same list of dataset types that are saved by PreExecInit and that should not be removed when we prune things. I think `_log` needs to be pruned if we want to re-run a task, otherwise we won't be able to store new log records from that task.

Reply:
Yes, if you are pruning then _log also needs to be pruned.